### PR TITLE
fix: Respect continueOn for leaf tasks

### PR DESCRIPTION
**Pull request by simster7**
Fixes: https://github.com/argoproj/argo/issues/4407

Signed-off-by: Simon Behar <simbeh7@gmail.com>

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: workflow/controller/dag_test.go [hunk 1]
```diff
@@ -2861,3 +2861,210 @@ func TestFailsWithParamDAG(t *testing.T) {
...
+    cpu: 13
+    memory: 6
+  startedAt: "2020-11-04T16:17:43Z"
+
+`
+
+func TestLeafContinueOn(t *testing.T) {
+	cancel, controller := newController()
+	defer cancel()
+	wfcset := controller.wfclientset.ArgoprojV1alpha1().Workflows("")
```
Comment (alexec) [inline workflow/controller/dag_test.go:3061]: minor:

```
wf := unmarshalWF(testLeafContinueOn)
cancel, controller := newController(wf)
```

File: workflow/controller/dag_test.go [hunk 2]
```diff
@@ -2861,3 +2861,210 @@ func TestFailsWithParamDAG(t *testing.T) {
 	woc.operate()
 	assert.Equal(t, wfv1.NodeFailed, woc.wf.Status.Phase)
 }
+
+var testLeafContinueOn = `
+apiVersion: argoproj.io/v1alpha1
```
Comment (alexec) [inline workflow/controller/dag_test.go:2866]: please minimize this YAML

**Review by alexec — CHANGES_REQUESTED** [review]

File: workflow/controller/dag_test.go (hunk 1 above)
Comment (simster7) [reply workflow/controller/dag_test.go]: Done

File: workflow/controller/dag_test.go (hunk 2 above)
Comment (simster7) [reply workflow/controller/dag_test.go:2866]: Done

**Comment by simster7** [comment]
@alexec May I get a re-review on this please?

File: workflow/controller/dag_test.go [hunk 3]
```diff
@@ -2861,3 +2861,127 @@ func TestFailsWithParamDAG(t *testing.T) {
...
+kind: Workflow
+metadata:
+  name: build-wf-kpxvm
+spec:
+  arguments: {}
```
Comment (alexec) [inline workflow/controller/dag_test.go:2871]: minor - please clean up this example - all the additional lines make it hard to understand which ones actually influence the test

**Review by alexec — APPROVED** [review]

File: workflow/controller/dag_test.go (hunk 3 above)
Comment (simster7) [reply workflow/controller/dag_test.go:2871]: Test is already minimized as per https://github.com/argoproj/argo/pull/4455#discussion_r517486368

Comment (alexec) [reply workflow/controller/dag_test.go:2871]: this is minor - but cant' you remove `arguments: {}` for example?

**Pull request merged**